In [10]:
from datascience import *
import numpy as np
from math import *

## Joint Distributions

Recall from Lessons 13 & 14: let $X$ be a random variable. $X$ has a distribution that is described by a probability mass function (pmf) or probability density function (pdf). 

We can consider multiple random variables simultaneously using joint distributions. 

NOTE: When answering the questions below, you are not required to use Python. If you would like to answer in Markdown, feel free to change the type of the cell, or to use both types of cells. 

#### Example 1: Discrete Joint Distribution

Let $X$ and $Y$ be discrete random variables that can each only take the values 0, 1 or 2, and do so according to the following distribution:


 |  | $X$  | 1 | 2 | 3 |
 |-- | ------ | ------ |
 |  | 1 | 0.17 | 0.15 | 0.08 |
 | $Y$ | 2 | 0.00 | 0.10 | 0.10 |
 |  | 3 | 0.08 | 0.20 | 0.12 |
 


This is an example of a joint probability mass function (joint pmf), and is denoted as $f_{X,Y}(x,y)$. 

For example, the probability that $X$ takes the value 1 AND $Y$ takes the value 3, or $P(X=1,Y=3)$ is equal to 0.08. 

**_1.1_** Find $P(X = 2,Y=3)$. 

**_1.2_** Find $E(X+Y)$. 

1.1 P(x=2,y=3) = 0.20


In [30]:
expected_val_x_y = 2*(.17) + 3*(.15) + 4*(.10+ .08+ .08) + 5*(.20 +.10) + 6*(.12)
expected_val_x_y

4.05

### Marginal Probability

When given a joint pmf like this, we may want to know the distribution of $X$ or $Y$ individually. Specifically, we might want to know $P(X=1)$ or $f_Y(y)$. 

Marginal probability can be found by summing across the remaining variable. Specifically,

$$
f_X(x)=\sum_y f_{X,Y}(x,y)
$$
and 
$$
f_Y(y)=\sum_x f_{X,Y}(x,y)
$$


**_1.3_** Find $P(X=2)$. 

**_1.4_** Find $f_Y(y)$ (the marginal pmf of $Y$). 

**_1.5_** Find $E(Y)$ and $V(Y)$. 

In [31]:
#1.3  
print('P(X=2) =', 0.45)

P(X=2) = 0.45


**1.4 : **  

In [32]:
import datascience as ds
marginal_pmf_y = ds.Table().with_columns('y', np.arange(3), 'probability', [0.40, 0.20, 0.40])
marginal_pmf_y

y    | probability
0    | 0.4
1    | 0.2
2    | 0.4

In [33]:
expected_y = 1*(.17+.15+.08) + 2*(.1 + .1) + 3*(.08+.20+.12)
print('E(Y) = ',expected_y)
variance_y = 1*(.17+.15+.08) + 4*(.1 + .1) + 9*(.08+.20+.12) - expected_y**2
print('V(Y) =',variance_y)

E(Y) =  2.0
V(Y) = 0.8000000000000007


### Conditional Probability

We may be interested in the probability $X$ takes a specific value conditioned on the value of $Y$. Recall that conditional probability is given by $P(A|B)=\frac{P(A,B)}{P(B)}$. 

So, essentially, conditional probability can be found by dividing the joint probability by the appropriate marginal probability. 

**_1.6_** Find $P(Y=1|X=3)$. 

**_1.7_** Find $f_{X|Y=2}(x)$, the conditional pmf of $X$, given $Y=2$. 

**_1.8_** Find $E(X|Y=2)$ and $V(X|Y=2)$. 

In [34]:
#1.6
prob_y_1_given_x_3 = .08/(.08+.12+.10)
print ('P(Y=1|X=3) =',prob_y_1_given_x_3 )

P(Y=1|X=3) = 0.2666666666666666


In [35]:
# 1.7
p_x_1_given_y_2 = 0/(.10+.10)
p_x_2_given_y_2 = .10/(.10+.10)
p_x_3_given_y_2 = .10/(.10+.10)
table_x_g_y = Table().with_columns('X', make_array(1,2,3), 'X Given Y =2', make_array(p_x_1_given_y_2, p_x_2_given_y_2, p_x_3_given_y_2))
print('Conditional pmf of X given Y = 2:')
table_x_g_y

Conditional pmf of X given Y = 2:


X    | X Given Y =2
1    | 0
2    | 0.5
3    | 0.5

In [36]:
# 1.8
expected_val_x_given_y_2 =  2*(.10/.2) + 3*(.10/.2)
varience_x_given_y_2 = (4*(.10/.2) + 9*(.10/.2)) - expected_val_x_given_y_2**2
print('expected value of x given y is 2:')
print(expected_val_x_given_y_2)
print('variance of x given y is 2: ')
print(varience_x_given_y_2)

expected value of x given y is 2:
2.5
variance of x given y is 2: 
0.25


**_1.9_** Are $X$ and $Y$ independent? Why or why not? 

No, they are dependent because when we did the test with the value of x = 1 and y = 3, it failed the interdependence test. The test determines if the P(x,y) = P(x) * P(y). This test failed, so they are dependent. 

### Covariance and Correlation

Expected value and variance help us characterize $X$ and $Y$ marginally and conditionally, but we may also be interested in measuring the relationship between $X$ and $Y$. For this, we use *covariance*. 

$$
Cov(X,Y)=E[(X-E(X))(Y-E(Y))] = E(XY)-E(X)E(Y)
$$

Note that if $X$ and $Y$ are independent, $Cov(X,Y) =0$. The converse is NOT necessarily true. 

Covariance is dependent on the scales of $X$ and $Y$, so if the two variables are of vastly different scale, we'll want to use covariance's unitless counterpart, correlation, denoted by $\rho$. 

$$
\rho = \frac{Cov(X,Y)}{\sqrt{Var(X)Var(Y)}}
$$

$\rho$ is bounded by the interval $[-1,1]$. When $\rho=1$, $X$ and $Y$ are perfectly positively correlated. Similarly, when $\rho=-1$, $X$ and $Y$ are perfectly negatively correlated. 

**_1.10_** Find $Cov(X,Y)$

**_1.11_** Find $Corr(X,Y)$, or $\rho$. 

In [37]:
# 1.10
e_xy = 1*(.17) + 2*(.15) + 3*(.08+.08) + 4*(.10) + 6*(.10+.20) + 9*(.12)
e_x = 1*(.17+.08) + 2*(.15+.10+.20) + 3*(.08+.10+.12)
e_y = 1*(.17+.15+.08) + 2*(.10+.10) + 3*(.08+.20+.12)
e_x_y = e_x*e_y
covar = e_xy - e_x_y
print('Cov(X,Y)=', covar)

Cov(X,Y)= 0.13000000000000078


In [49]:
#1.11
variance_y = 1*(.17+.15+.08) + 4*(.1 + .1) + 9*(.08+.20+.12) - e_y**2
variance_x = 1*(.17+.08) + 4*(.15+.10+.20) + 9*(.08+.10+.12) - e_x**2
roe_denom = (variance_x*variance_y) **.5
roe_denom
roe = covar/roe_denom
print('Corr(X,Y) = ', roe)

Corr(X,Y) =  0.19642931269503836


In problem 1.2, we found $E(X+Y)$. In order to find $Var(X+Y)$, we need to know how $X$ and $Y$ are correlated: 

$$
Var(X+Y)= Var(X)+Var(Y)+2*Cov(X,Y)
$$

**_1.12_** Find $Var(X+Y)$. 

In [39]:
var_x_y = variance_x + variance_y + 2*covar
var_x_y

1.6075000000000026

#### Example 2: Continuous Joint Distribution

All of the concepts above apply to continuous random variables. Consider continuous random variables $X$ and $Y$ with the following joint pdf:

$$
f_{X,Y}(x,y)=k(x+y)
$$

where both $x$ and $y$ are bounded by the interval $[0,1]$. 

**_2.1_** Find the value of $k$ that makes $f$ a valid joint pdf. 

In [40]:
# k would take a value of 1. This is because in order for it to be a valid pmf, 
# the integral of that function bound by [0,1] has to equal 1. k with a value of 
# 1 is the only value that makes this true since the function integrates to 1 without the k value. 
from scipy import integrate
without_k = np.round(integrate.dblquad(lambda x,y:(x + y),0,1, 0, 1)[0],10)
without_k

1.0

**_2.2_** Find $P(X<0.5,Y<0.5)$

In [41]:
from scipy import integrate
np.round(integrate.dblquad(lambda x,y:(x+y), 0, .5, 0, .5)[0], 10)


0.125

**the proability for this function is 0.125**


**_2.3_** Find $f_X(x)$ and $f_Y(y)$, the marginal pdfs of $X$ and $Y$. 

**_2.4_** Find $E(X)$ and $E(Y)$. 

In [51]:
#2.3

print('the marginal pdf of x is: x + .5')
print('the marginal pdf of y is y + .5')

print('to find these values, I integrated the function with respect to the opposite varaible. I could not figure out how to do this in matlab where it is able to return a variable')

the marginal pdf of x is: x + .5
the marginal pdf of y is y + .5
to find these values, I integrated the function with respect to the opposite varaible. I could not figure out how to do this in matlab where it is able to return a variable


In [42]:
expected_x = np.round(integrate.quad(lambda x:x*(x + .5), 0, 1)[0],10)
expected_x

0.5833333333

In [43]:
expected_y =  np.round(integrate.quad(lambda y:y*(y + .5), 0, 1)[0],10)
expected_y

0.5833333333

**_2.5_** Find $P(X>0.5\vert Y\leq 0.5)$. 

In [46]:
#2.5 probability that x>.5 given y</ .5
prob_yless_5 = integrate.dblquad(lambda x, y: 1*(x+y), 0, .5, lambda x: 0, lambda x: 1)[0]
prob_total = integrate.dblquad(lambda x,y:1*(x+y), 0, .5, lambda x: .5, lambda x: 1)[0]
prob_final = prob_total/prob_yless_5
prob_final

0.6666666666666666

**_2.6_** Find the conditional distributions of $X|Y$ and $Y|X$. Recall that conditional distributions can be found by dividing the joint pdf by the relevant marginal pdf. 

**_2.7_** Find $E(X|Y)$ and $E(Y|X)$.

In [62]:
#2.6
def XgivenY(x,y):
    return (x+y)/.5+y
def YgivenX(x,y):
    return (x+y)/.5+x
print('for this i just took the joint pdf found above and divided it by the marginal pdf also above')
print('fx|y(X|Y) = (x+y)/(y+.5)')
print('fy|x(Y|X) = (x+y)/(x+.5)')

for this i just took the joint pdf found above and divided it by the marginal pdf also above
fx|y(X|Y) = (x+y)/(y+.5)
fy|x(Y|X) = (x+y)/(x+.5)


In [63]:
#2.7 expected values through integration

EXgivenY = integrate.dblquad(lambda x,y:XgivenY(x,y)*x, 0, 1, lambda x:0, lambda x:1)[0]
EYgivenX =  integrate.dblquad(lambda x,y:YgivenX(x,y)*y, 0, 1, lambda x:0, lambda x:1)[0]
print('E(X|Y)=', EXgivenY, 'E(Y|X) =', EYgivenX)

E(X|Y)= 1.4166666666666665 E(Y|X) = 1.4166666666666665


**_2.8_** Are $X$ and $Y$ independent? 

In [67]:
#2.8
print('in order for them to be independent, expected value of x would have to equal expected value of x given y:')
print(expected_x == EXgivenY)
print('since this output is false, X and Y are not independent')

in order for them to be independent, expected value of x would have to equal expected value of x given y:
False
since this output is false, X and Y are not independent


**_2.9_** What is $Cov(X,Y)$? 

In [71]:
#2.9
CovXY = integrate.dblquad(lambda x, y: (x+y)-(.5+x)*(.5-y), 0, 1, lambda x: 0, lambda x: 1)[0]
CovXY

0.9999999999999999